### DISCUSSION TITLE
Processing of mbed_app.json

### DISCUSSION THREAD

**Discussion by JojoS62** [mbed-ce/mbed-os#86] [General]
When is mbed_app.json processed? I have a project and added ```"target.printf_lib": "std",```, but the change was not not detected, also not when I selected a variant again to trigger a new cmake configuration.
I had to delete the build dir and recompile everything again. Is there some other way to trigger this configuration step?

Which parts are compatible with mbed-os build system? Does this have an effect on the build: ```"device_has_remove": ["SLEEP"]``` ?

**Comment by JohnK1987** [comment]
I had same feeling, but it seems CMake does this correctly.
When I change something in `mbed_app.json` and do variant re-selecting, then message below will be in output console. 
```
[cmake] -- Mbed: mbed_app.json modified, regenerating configs...
```

BR, Jan

**Comment by multiplemonomials** [comment]
mbed_app.json is processed by [this CMake script](https://github.com/mbed-ce/mbed-os/blob/master/tools/cmake/mbed_generate_configuration.cmake).  It watches the json files for changes, and then reruns the Mbed Tools package to generate the configuration whenever the json file is modified.

When you modify mbed_app.json, then rerun CMake, it should notice the changes and reconfigure (the previous modification time is stored as a CMake cache variable.  I have seen this work in the past, if it didn't for you, that's a bug.

If you want to manually trigger the configuration, that's easy.  You just have to delete the `<build dir>/mbed_config.cmake` file.  CMake will notice this and reconfigure everything.

**Comment by multiplemonomials** [comment]
> Which parts are compatible with mbed-os build system? Does this have an effect on the build: "device_has_remove": ["SLEEP"] ?

Yes, that does have an effect on CMake.  It alters the `MBED_TARGET_DEFINITIONS` list (defined in the generated mbed_config.cmake file) to remove the entry `DEVICE_SLEEP=1`.  It also removes the `#define DEVICE_SLEEP 1` compile definition.  So, CMake code and/or C++ code can do different things based on that line.

**Comment by JojoS62** [comment]
selecting a variant works to get modifications in mbed_app.json, thanks for hints.

I've found this about adding depencies to rerun cmake, would that work with the VSCode extension? Where should it be placed?


```cmake
function(watch)
    set_property(
            DIRECTORY
            APPEND
            PROPERTY CMAKE_CONFIGURE_DEPENDS ${ARGV}
    )
endfunction()
```

https://www.reddit.com/r/cmake/comments/iokem9/force_cmake_rerun_when_file_changes/

**Reply by multiplemonomials** [reply]
Ooh interesting, I didn't know about that.  It should go in mbed_generate_configuration.cmake.
